src/bufr/DataObject.h

```diff
@@ -0,0 +1,368 @@
+        /// \param compressionLevel The GZip compression level to use, must be 0-9
+        virtual ioda::Variable createVariable(
+                                      ioda::ObsGroup& obsGroup,
+                                      const std::string& name,
+                                      const std::vector<ioda::Variable>& dimensions,
+                                      const std::vector<ioda::Dimensions_t>& chunks,
+                                      int compressionLevel) const = 0;
+
+        /// \brief Slice the data object given a vector of row indices.
+        /// \param slice The indices to slice.
+        /// \return Slice of the data object.
+        virtual std::shared_ptr<DataObjectBase>
+            slice(const std::vector<std::size_t>& rows) const = 0;
+
+     protected:
+        std::string fieldName_;
+        std::string groupByFieldName_;
+        Dimensions dims_;
+        std::string query_;
+        std::vector<std::string> dimPaths_;
+    };
+
+
+    template <typename T>
+    class DataObject : public DataObjectBase
+    {
+     public:
+        typedef T value_type;
+
+        /// \brief Constructor.
+        /// \param dimensions The dimensions of the data object.
+        DataObject(const std::vector<T>& data,
+                   const std::string& field_name,
+                   const std::string& group_by_field_name,
+                   const Dimensions& dimensions,
+                   const std::string& query,
+                   const std::vector<std::string>& dimPaths) :
+            DataObjectBase(field_name, group_by_field_name, dimensions, query, dimPaths),
+            data_(data)
+        {};
+
+        ~DataObject() = default;
+
+        /// \brief Makes an ioda::Variable and ads it to the given ioda::ObsGroup
+        /// \param obsGroup Obsgroup were to add the variable
+        /// \param name The name to associate with the variable (ex "latitude@MetaData")
+        /// \param dimensions List of Variables to use as the dimensions for this new variable
+        /// \param chunks List of integers specifying the chunking dimensions
+        /// \param compressionLevel The GZip compression level to use, must be 0-9
+        ioda::Variable createVariable(ioda::ObsGroup& obsGroup,
+                                      const std::string& name,
+                                      const std::vector<ioda::Variable>& dimensions,
+                                      const std::vector<ioda::Dimensions_t>& chunks,
+                                      int compressionLevel) const final
+        {
+            auto params = makeCreationParams(chunks, compressionLevel);
+            auto var = obsGroup.vars.createWithScales<T>(name, dimensions, params);
+            var.write(data_);
+            return var;
+        };
+
+        /// \brief Print data to stdout for debug purposes.
+        void print(std::ostream &out) const final
```

Comment:
Is this needed still? Should it use OOPS if so?

Reply:
The objective was to print the contents of the data object out for debug purposes. I don't think its currently used, just a convenience method for debugging.

Reply:
the argument lets you print the data to any stream you want. So you could direct it to the oops log stream if you desired.